### DISCUSSION TITLE
Foreign keys of viewonly relationship?

### DISCUSSION THREAD

**Discussion by asbuch** [sqlalchemy/sqlalchemy#12005] [Usage Questions]
Is the `foreign_keys` parameter used for anything in a relationship defined with `viewonly=True`?

The following code results in an error if the line `foreign_keys=parent_id` is deleted. Is that because the `foreign_keys` parameter is used for something, or because `viewonly` is only checked at the time where sqlalchemy is about to change the foreign key column?

```
class Person(Base):
    __tablename__ = 'people'
    id = Column(Integer, primary_key=True)
    parent_id = Column(Integer)
    children = relationship(
        'Person',
        primaryjoin=id == remote(parent_id),
        viewonly=True,
        foreign_keys=parent_id
    )
```

**Comment by CaselIT** [comment]
Hi,

viewonly just disables the synchronization of the other side of the relationship; everything else works the same

**Reply by asbuch** [reply]
What I am hoping to understand is what "everything else" is! When `primaryjoin` is specified, my understanding is that `foreign_keys` is only used to determine which columns to set when synchronizing. So if synchronization is disabled, then the value of `foreign_keys` should not matter?

In the following enlarged example, does it make any difference if I use `foreign_keys=parent_id1` or `foreign_keys=(parent_id1, parent_id2)` or `foreign_keys=(parent_id2, unrelated)`?

```
class Person(Base):
    __tablename__ = 'people'
    id1 = Column(Integer, primary_key=True)
    id2 = Column(Integer, primary_key=True)
    parent_id1 = Column(Integer)
    parent_id2 = Column(Integer)
    unrelated = Column(Integer)
    children = relationship(
        'Person',
        primaryjoin=and_(id1 == remote(parent_id1), id2 == remote(parent_id2)),
        viewonly=True,
        # foreign_keys=parent_id1  # works
        # foreign_keys=(parent_id1, parent_id2)  # works
        foreign_keys=(parent_id2, unrelated)  # works
        # foreign_keys=(),  # gives error
        # foreign_keys=unrelated  # gives error
    )
```

**Reply by CaselIT** [reply]
> So if synchronization is disabled, then the value of foreign_keys should not matter?

Viewonly does not disable that logic of relationships, that still figures out what it should do. It literally only disable that actual update that would happen normally
